src/server/db/migrations/2020-11-13_0.3.0_to_0.4.0/migrate_0.3.0_to_0.4.0.js

```diff
@@ -0,0 +1,414 @@
+ *
+ * @param {String} responseString - the response from the request to get all documents
+ * @param {Object} options - Where to store the to-be-updated documents
+ *
+ * @return {Array} - The updated documents in the new data structure with the new schema version.
+ */
+sjrk.storyUpdate.updateStoriesData = function (responseString, options) {
+    var allDocs = JSON.parse(responseString);
+    var updatedDocs = [];
+    options.totalNumOfDocs = allDocs.total_rows;
+
+    if (allDocs.rows) {
+        fluid.each(allDocs.rows, function (aRow) {
+            var aDoc = aRow.doc;
+            // To filter out the "_design/views" doc that doesn't have the "schemaVersion" field
+            if (aDoc.type === "story") {
+                // Make the required updates, adding missing fields
+                // Please consult the list at the top of this file for details on each item
+
+                if (!aDoc.schemaVersion || aDoc.schemaVersion < sjrk.storyUpdate.newSchemaVersion) {
+
+                    aDoc.schemaVersion = sjrk.storyUpdate.newSchemaVersion; // 2.a
+
+                    aDoc.value.id = aDoc.value.id || aDoc._id; // 2.b
+                    aDoc.value.published = aDoc.value.published || true; // 2.c
+                    aDoc.value.timestampCreated = aDoc.value.timestampCreated || new Date().toDateString(); // 2.d
+                    aDoc.value.timestampPublished = aDoc.value.timestampPublished || new Date().toDateString(); // 2.e
+
+                    var blocks = fluid.copy(aDoc.value.content);
+
+                    fluid.each(blocks, function (block, i) {
+                        block.order = block.order || i; // 2.f
+                        block.firstInOrder = block.firstInOrder || i === 0 ? true : false; // 2.g
+                        block.lastInOrder = block.lastInOrder || i === aDoc.value.content.length - 1 ? true : false; // 2.h
+
+                        if (block.imageUrl) {
+                            block.mediaUrl = block.imageUrl; // 3
+                        }
+
+                        if (block.mediaUrl && !block.mediaUrl.startsWith(sjrk.storyUpdate.uploadsDirectoryPrefix)) {
+                            block.mediaUrl = sjrk.storyUpdate.uploadsDirectoryPrefix + block.mediaUrl; // 6
+                        }
+
+                        // remove obsolete/unused block keys, steps 3 & 4
+                        block = fluid.censorKeys(block, sjrk.storyUpdate.blockValuesToRemove);
+
+                        aDoc.value.content[i] = block;
+                    });
+
+                    // remove obsolete/unused story keys, step 5
+                    aDoc.value = fluid.censorKeys(aDoc.value, sjrk.storyUpdate.storyValuesToRemove);
+
+                    fluid.log("Updating the story with ID: ", aDoc._id);
+                    updatedDocs.push(aDoc);
+                }
+            }
+        });
+
+        options.updatedDocs = updatedDocs;
+    }
+
+    return updatedDocs;
```

Comment:
Given the return values of all these functions are ignored by the caller, it seems odd to return them.
You could either axe the returns, or else turn this into a slightly more "functional" pipeline involving fluid.promise.fireTransformEvent - this seems to make sense since the yielded value from sjrk.storyUpdate.updateStoriesData is actually supplied to sjrk.storyUpdate.updateDB - the latter can then get the signature sjrk.storyUpdate.updateDB(updatedDocs, options). "responseString" then becomes the previous payload in the chain.
This would also result in some simplification of the logic in gpii.dbRequest.createResponseHandler which accepts a function returning a promise, and then accepts another promise - it could then simply return a promise for its yielded result.
This might be a too ambitious refactoring for the current situation but we should leave the recommendation as a note in case this code gets revisited (I see it has also hopped across from another project once).

Reply:
This is an excellent idea. I have captured your feedback in [SJRK-429](https://issues.fluidproject.org/browse/SJRK-429) and have added a note above that line in the code.